lib_eio/net.mli

```diff
@@ -113,7 +115,32 @@ end
 val connect : sw:Switch.t -> #t -> Sockaddr.stream -> <stream_socket; Flow.close>
 (** [connect ~sw t addr] is a new socket connected to remote address [addr].
 
-    The new socket will be closed when [sw] finishes, unless closed manually first. *)
+    The new socket will be closed when [sw] finishes, unless closed manually first.
+
+    @raise Unix_error if connection couldn't be established. *)
+
+val with_tcp_connect :
+  ?timeout:float ->
+  host:string ->
+  service:string ->
+  <clock: #Time.clock; net: #t; ..> ->
```

Comment:
We should probably define `type timeout = Clock.t * float` or similar, and then do:
```suggestion
val with_tcp_connect :
  ?timeout:Timeout.t ->
  t ->
  host:string ->
  service:string ->
```

Reply:
If we make clock part of `timeout` and then make `timeout` an optional parameter, does this mean we don't want connection timeout? i.e. if a user didn't use `timeout` parameter, we have a default value of `0.1` for timeout seconds but what do we use as default clock value?

If no timeout is used while connecting to the a host, wouldn't that degrade user experience of the API?

Reply:
Generally I prefer to avoid default timeouts and let the calling code handle that. We have no idea whether the network will require more or less than 0.1s to make a connection.

There's also the [Happy Eyeballs](https://datatracker.ietf.org/doc/html/rfc8305) protocol, which tries IPv4 and IPv6 in parallel. However, user applications can use a library (e.g. https://github.com/roburio/happy-eyeballs) for that.

Reply:
>There's also the [Happy Eyeballs](https://datatracker.ietf.org/doc/html/rfc8305) protocol, which tries IPv4 and >IPv6 in parallel. However, user applications can use a library (e.g. https://github.com/roburio/happy-eyeballs) >for that.

Yes, indeed this PR was in part influenced by the Happy Eyeballs RFC in terms of IP version selection and the timeout requirements.

Reply:
Done.